tripal/tripal
# Drush command to update chado instance to match Tripal Term expectations

# Tripal 4 Core Dev Task

### Issue #1698 


### Tripal Version: 4

## Description


This PR adds a drush command that will check a specific chado instance to ensure it matches the expectations set out by Tripal 4. This is needed because chado vocabularies and cvterms often become a bit eccentric over the years ;-p Right now, we are finding that running the prepare step in Tripal 4 on a chado exported from a Tripal 3 site fails due to changes in term expectations, definitions, etc.

More specifically, the drush command `drush trp-check-terms --chado_schema=chado` would run a number of checks on the cv, db, cvterm, and dbxref tables to ensure the values associated with key Tripal terms match those defined in the [Chado Content Terms YAML](https://github.com/tripal/tripal/blob/4.x/tripal_chado/config/install/tripal.tripal_content_terms.chado_content_terms.yml). This command also offers the option in some cases to fix any deviations from what is expected.

Specific cases which are checked:
- checks if the vocabulary exists in the cv table where cv.name = vocab.label
  - if the cv exists then compares the definition and warns if they differ
- checks if the ID Spaces for this vocabulary exist in the db table where db.name = idSpace.name
  - if the db exists then compares the description, urlprefix and url and warns if they differ.
- checks if the term exists as a unique cvterm/dbxref combo and reports success if yes.
  - reports errors for terms if:

    1. cvterm exists (cv.name = vocab.name + cvterm.name = term.name) but is attached to the wrong dbxref.
    2. cvterm exists with the term name but wrong cv; however, it is connected to the right dbxref
    3. dbxref exists but is connected to 1+ cvterms that do not match the term we are looking for.
    4. cvterm exists but is connected to a dbxref where the accession matches but it's in a different id space.

  - warns if the cvterm definition doesn't match what we expected.
  - reports but not an error if:

    1. dbxref exists but it is not connected to any cvterms
    2. neither cvterm nor dbxref exist
 - checks if the same term is defined in YAML configuration more than once

## Testing?




I started automated testing for this!

### Manual Testing via Docker

1. Start a fresh docker on this branch.
```bash
git clone https://github.com/tripal/tripal trpCore-1895-checkTerms
cd trpCore-1895-checkTerms
git checkout tv4g2-issue1698-unable-to-prepare-an-existing-chado-database-from-tripal-3
docker build --tag=tripaldocker:tripal1698 --build-arg drupalversion=10.2.x-dev --build-arg postgresqlversion=16 --file tripaldocker/Dockerfile-php8.3 ./
docker run --publish=80:80 -tid --name=tripal1698 --volume=/Users/las166/Docker/trpCore-1698-prepareChado:/var/www/drupal/web/modules/contrib/tripal tripaldocker:tripal1698
docker exec tripal1698 service postgresql restart
```
2. Run the drush command on the freshly installed chado. We expect to see the following output which indicates there are no errors and everything matches our expectations. There should not be any red/yellow in the summary table.
<img width="1163" alt="Screenshot 2024-06-17 at 3 15 23 PM" src="https://github.com/tripal/tripal/assets/1566301/799ad645-37c5-4326-a10e-b0aa4ad334bb">
<img width="1165" alt="Screenshot 2024-06-17 at 3 15 40 PM" src="https://github.com/tripal/tripal/assets/1566301/a9cb61cc-b4f7-4a42-8d93-1ab7ba3c9258">
4. Now you want to make some changes which will cause warnings:

```sql
UPDATE chado.cv SET definition='CHANGED FOR TESTING' WHERE cv.name = 'hydra';
UPDATE chado.db SET description='CHANGED FOR TESTING' WHERE db.name='schema'; 
UPDATE chado.db SET url='not/a/url', urlprefix='something/completely/different?{db}:{accession}'  WHERE db.name='TAXRANK';
UPDATE chado.cvterm SET definition='CHANGED FOR TESTING' WHERE name='Physical map';
```
5. Now run the drush command and check that you get warnings for all terms that are in the hydra vocabulary or the schema id space. There should also be a warning for the 'Physical map' term.
<img width="1166" alt="Screenshot 2024-06-17 at 3 24 43 PM" src="https://github.com/tripal/tripal/assets/1566301/620f5f7b-f90d-43cb-a7a6-dcd3bcd44eb7">
<img width="1166" alt="Screenshot 2024-06-17 at 3 25 00 PM" src="https://github.com/tripal/tripal/assets/1566301/1ad0705d-a7d7-4cb7-a294-d4598b04ce1d">
It should prompt asking if you want to see a summary of the terms, choose yes and confirm you see the following summaries. After each summary it also asks if you want us to fix it. Also select yes.
<img width="1166" alt="Screenshot 2024-06-17 at 3 27 22 PM" src="https://github.com/tripal/tripal/assets/1566301/10774b55-01b9-41b1-aafa-fca9845cb04e">
6. Now run the command again and confirm the warnings no longer show up. You can also query the database to ensure the fix is done right.
7. Now we want to trigger some of the errors. This can be done with the following commands:

 - Correct cvterm, dbxref found but db is wrong
  ```sql
  INSERT INTO chado.db (name) VALUES ('DATA');
  UPDATE chado.dbxref SET db_id=42 WHERE dbxref_id=8;
  ```
 - cvterm exists, dbxref exists but connection is broken (i.e. cvterm connected to different dbxref).
  ```sql
  INSERT INTO chado.dbxref (db_id, accession) VALUES (42, 'NOTREAL');
  UPDATE chado.cvterm SET dbxref_id=3496 WHERE dbxref_id=13;
  ```
 - cvterm exists but is connected to wrong dbxref and the correct one doesn't exist.
  ```sql
  UPDATE chado.dbxref SET accession='UPDATEEXISTING', db_id=42 WHERE dbxref_id=24;
  ```
 - cvterm doesn't exist but dbxref does and it's connected to the wrong cvterm
  ```sql
  DELETE FROM chado.cvterm WHERE cvterm_id=215;
  UPDATE chado.cvterm SET dbxref_id=215 WHERE cvterm_id=2133;
  ```
 - cvterm with that name exists and is attached to the right dbxref but cv is wrong
  ```sql
  INSERT INTO chado.cv (name) VALUES ('NEWCV');
  UPDATE chado.cvterm SET cv_id=34 WHERE cvterm_id=219;
  ```
8. Now run the command and make sure it catches all these problems. 

```bash
docker exec -it tripal1698 drush trp-check-terms --chado_schema=chado
```
<img width="1159" alt="Screenshot 2024-06-17 at 5 07 46 PM" src="https://github.com/tripal/tripal/assets/1566301/2c43f45c-3b32-4ba4-adbc-9cd97660c2a0">
<img width="1166" alt="Screenshot 2024-06-17 at 5 09 36 PM" src="https://github.com/tripal/tripal/assets/1566301/cba421fa-840f-4323-a8c3-ab742a31521f">
